src/ol-ext/format/featurehash.js

```diff
@@ -0,0 +1,940 @@
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writeMultiPointGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPoint);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var end = flatCoordinates.length;
+  return 'P(' + this.encodeCoordinates_(flatCoordinates, stride, 0, end) + ')';
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @return {string} Encoded geometry.
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writePolygonGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.Polygon);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var ends = geometry.getEnds();
+  var linearRingCount = ends.length;
+  var offset = 0;
+  var textArray = ['a('];
+  for (var i = 0; i < linearRingCount; ++i) {
+    // skip the "closing" point
+    var end = ends[i] - stride;
+    var text = this.encodeCoordinates_(flatCoordinates, stride, offset, end);
+    if (i !== 0) {
+      textArray.push('\'');
+    }
+    textArray.push(text);
+    offset = ends[i];
+  }
+  textArray.push(')');
+  return textArray.join('');
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @return {string} Encoded geometry.
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writeMultiPolygonGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPolygon);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var endss = geometry.getEndss();
+  var polygonCount = endss.length;
+  var offset = 0;
+  var textArray = ['A'];
+  for (var i = 0; i < polygonCount; ++i) {
+    textArray.push('(');
+    var ends = endss[i];
+    var linearRingCount = ends.length;
+    for (var j = 0; j < linearRingCount; ++j) {
+      // skip the "closing" point
+      var end = ends[j] - stride;
+      var text = this.encodeCoordinates_(flatCoordinates, stride, offset, end);
+      if (j !== 0) {
+        textArray.push('\'');
+      }
+      textArray.push(text);
+      offset = ends[j];
+    }
+    textArray.push(')');
+  }
+  return textArray.join('');
+};
+
+
+/**
+ * @const
+ * @private
+ * @type {Object.<string, function(string, olx.format.ReadOptions=):ol.geom.Geometry>}
+ */
+ngeo.format.FeatureHash.GEOMETRY_READERS_ = {
+  'P': ngeo.format.FeatureHash.readMultiPointGeometry_,
+  'L': ngeo.format.FeatureHash.readMultiLineStringGeometry_,
+  'A': ngeo.format.FeatureHash.readMultiPolygonGeometry_,
+  'l': ngeo.format.FeatureHash.readLineStringGeometry_,
+  'p': ngeo.format.FeatureHash.readPointGeometry_,
+  'a': ngeo.format.FeatureHash.readPolygonGeometry_
+};
```

Comment:
I haven't taken this comment into account, as I don't quite see how I could address it in practice.